File: Reverse_Linked_List.cpp

#include <stdio.h>
#include <stdlib.h>
#include<iostream>
using namespace std;


struct Node {
    int data;
    struct Node *next;
    Node(int x)
    {
        data = x;
        next = NULL;
    }
};

class Solution
{
    public:
    struct Node* reverseList(struct Node *head)
    {
        Node *prev=NULL;
        Node*curr=head;
        while(curr){
            Node *forward=curr->next;
            curr->next=prev;
            prev=curr;
            curr=forward;
        }
        return prev;
    }
    
};
    



void printList(struct Node *head)
{
    struct Node *temp = head;
    while (temp != NULL)
    {
       printf("%d ", temp->data);
       temp  = temp->next;
    }
}

int main()
{
    int T,n,l,firstdata;
    cin>>T;

    while(T--)
    {
        struct Node *head = NULL,  *tail = NULL;

        cin>>n;
        
        cin>>firstdata;
        head = new Node(firstdata);
        tail = head;
        
        for (int i=1; i<n; i++)
        {
            cin>>l;
            tail->next = new Node(l);
            tail = tail->next;
        }
        
        Solution ob;
        head = ob. reverseList(head);
        
        printList(head);
        cout << endl;
    }
    return 0;
}

File: reverse_linkedlist.java
class ListNode {
    int val;
    ListNode next;
    ListNode() {}
    ListNode(int val) { 
        this.val = val; 
        this.next=null;
    }
    ListNode(int val, ListNode next) { this.val = val; this.next = next; }
}

public class Main
{
    public ListNode reverseList(ListNode head) {
        ListNode temp,temp1,temp2;
        temp=head;
        if(head==null || head.next==null){
            return head;
        }
        temp1=head.next;
        head.next=null;
        while(temp1.next!=null){
            temp2=temp1.next;
            temp1.next=temp;
            temp=temp1;
            temp1=temp2;
        }
        temp1.next=temp;
        return temp1;

    }
	public static void main(String[] args) {
		ListNode head=new ListNode(25);
        head.next=new ListNode(9);
        ListNode temp=head.next;
        temp.next=new ListNode(96);
        temp=temp.next;
        temp.next=new ListNode(75);
        ListNode temp1=head;
        while(temp1!=null){
            System.out.print(temp1.val+" ");
            temp1=temp1.next;
        }
        System.out.println();
        Main obj=new Main();
        temp1=obj.reverseList(head);
        while(temp1!=null){
            System.out.print(temp1.val+" ");
            temp1=temp1.next;
        }
        
	}
}
